### ISSUE TITLE
TRanform coordinates are not applied in the dock goal

### ISSUE THREAD

**Issue by TheConstructAi** [open-rmf/fleet_adapter_template#19]
Hi I'm having issues with using the original template when executing the docking method; when completing the docking, for some reason, the robots are sent to the docking node with coordinates that are not transformed.
The transformations between simulation and rmf coords work, except for the docking nav goals.

Any ideas why?

**Comment by Yadunund** [comment]
I'm not sure I understand the question fully. Both `follow_new_path()` and `dock()` methods need to be implemented by the user. In this template we provide most of the boiler plate needed. If your robot's dock API needs coordinates in your robot's map's frame, you should be able to do the transform and call you API. There is a block [here](https://github.com/open-rmf/fleet_adapter_template/blob/1fe945a20fcd6424b71f1c8c20600e7e4324c77d/fleet_adapter_template/fleet_adapter_template/RobotCommandHandle.py#L319-L322) that you need to replace with whatever logic you need.

**Comment by TheConstructAi** [comment]
Ok let me elaborate a bit more showing an example:

This is the code of _dock():

```
def dock(
            self,
            dock_name,
            docking_finished_callback):
        ''' Docking is very specific to each application. Hence, the user will
            need to customize this function accordingly. In this example, we
            assume the dock_name is the same as the name of the waypoints that
            the robot is trying to dock into. We then call api.start_process()
            to initiate the robot specific process. This could be to start a
            cleaning process or load/unload a cart for delivery.
        '''

        self._quit_dock_event.clear()
        if self._dock_thread is not None:
            self._dock_thread.join()

        self.dock_name = dock_name
        assert docking_finished_callback is not None
        self.docking_finished_callback = docking_finished_callback

        # Get the waypoint that the robot is trying to dock into
        dock_waypoint = self.graph.find_waypoint(self.dock_name)
        assert(dock_waypoint)
        self.dock_waypoint_index = dock_waypoint.index

        def _dock():
            # Request the robot to start the relevant process
            self.node.get_logger().info(
                f"Requesting robot {self.name} to dock at {self.dock_name}")
            self.api.start_process(self.name, self.dock_name, self.map_name)

            with self._lock:
                self.on_waypoint = None
                self.on_lane = None
            self.sleep_for(0.1)

            while (not self.api.process_completed(self.name)):
                # Check if we need to abort
                if self._quit_dock_event.is_set():
                    self.node.get_logger().info("Aborting docking")
                    return
                self.node.get_logger().info("Robot is processing Task...")
                self.sleep_for(0.1)

            with self._lock:
                self.on_waypoint = self.dock_waypoint_index
                self.dock_waypoint_index = None
                self.docking_finished_callback()
                self.node.get_logger().info(">>>>>>> Task completed")

        self._dock_thread = threading.Thread(target=_dock)
        self._dock_thread.start()
```

As you can see its the bilerplate, nothing else. So in theory it should dock to the docking point, right?
So now we send this task, were the last node is the docking point:

![docktest1](https://user-images.githubusercontent.com/80715513/235587051-1a983f4b-f7c7-4429-b0e0-5dd268c2264f.png)

When we sen it here it navigates relatively well, but when it has to dock the current robot pose just goes nuts, and it move the robot elsewhere. Then after a while, when the 30 seconds stay in place system comes into place , the robot, reaches that docking position. It really weird. The version we are using of rmf is the main branch


![issue](https://user-images.githubusercontent.com/80715513/235588192-e3a67a35-cfe3-40ac-89fe-aa123bdaa96a.gif)




 I have made the simulation 1 to 1 with the rmf map now, so it seems unlikely that is something related to the transformations done in the script. So my question is...What is it causing this behavior? 
It happens most of the times when docking while navigating to either points tends to be alright.

**Comment by Yadunund** [comment]
> As you can see its the boilerplate, nothing else. So, in theory it should dock to the docking point, right?

The problem is because you have not implemented the `_dock()` function by replacing the `# Implement your code here` with any actual code that performs docking. You fleet adapter is not doing anything but wrongly updates the state of the robot before calling `dock_finished_cb()`.

The way docking works is that when the robot is about to enter a lane where the exit vertex of the lane has a `dock_name` property, the fleet adapter will trigger the `RobotCommandHandle::dock(~)` function. The idea of this template is for you to replace the `# Implement your code here` with your own logic to get your robot to dock, ie, whatever logic you need to move your robot to the exit vertex. But since you have not added any logic, the `_dock()` will exit after updating the [internal state trackers](https://github.com/open-rmf/fleet_adapter_template/blob/1fe945a20fcd6424b71f1c8c20600e7e4324c77d/fleet_adapter_template/fleet_adapter_template/RobotCommandHandle.py#L332-L333) by effectively saying "your robot has reached the exit vertex". Then depending on when the next update_state() is triggered, it may either call [update_off_grid()](https://github.com/open-rmf/fleet_adapter_template/blob/1fe945a20fcd6424b71f1c8c20600e7e4324c77d/fleet_adapter_template/fleet_adapter_template/RobotCommandHandle.py#L416-L418) if `dock_completed_cb()` is not yet called or [update_current_waypoint()](https://github.com/open-rmf/fleet_adapter_template/blob/1fe945a20fcd6424b71f1c8c20600e7e4324c77d/fleet_adapter_template/fleet_adapter_template/RobotCommandHandle.py#L398-L400) which explains why you don't see repeatable behavior.

So then there is a problem where you're telling RMF your robot is at this exit vertex, but it actually is not. I'm not sure how you're controlling your robots in simulation but if you're using the `Fleet Manager` from `rmf_demos_fleet_adapter` then when your robot receives a navigation request, the first trajectory point will be the robot's current location which may be too far from the actual robot -> this could cause the `slotcar` plugin to misbehave.

TLDR; you should implement the `_dock()` function to move your robot to the exit vertext of the lane, ie the waypoint with the dock_name property.

**Comment by Yadunund** [comment]
You can reference our `rmf_demos_fleet_adapter` to see how we implement the `_dock()` function : https://github.com/open-rmf/rmf_demos/blob/3f29a858ae64f2702696c5b1e5481c969c8f08a4/rmf_demos_fleet_adapter/rmf_demos_fleet_adapter/RobotCommandHandle.py#L415-L474

**Comment by TheConstructAi** [comment]
Wow thanks a lot! All this info is handy! We will dig info it right away ;) Thanks @Yadunund 
Yeah its seems like the slot car is not prepared for that and does really strange things. The robots are just using plain old slotcar.
I'll post here the findings ;).
